compositor_pipeline/src/audio_mixer/internal_audio_mixer.rs

```diff
@@ -189,4 +73,200 @@ impl InternalAudioMixer {
+                    return;
+                }
+            }
+            None => self.last_enqueued_pts = Some(input_samples.start_pts),
+        }
+
+        let expected_samples_before = input_samples
+            .start_pts
+            .saturating_sub(start_pts)
+            .as_secs_f64()
+            * output_sample_rate as f64;
+        let missing_samples = (expected_samples_before as u64)
+            .saturating_sub(self.popped_samples)
+            .saturating_sub(self.samples.len() as u64);
+        if missing_samples > 10 {
+            for _ in 0..missing_samples {
+                self.samples.push_back((0, 0));
+            }
+        }
+
+        self.samples.extend(input_samples.samples.iter());
+    }
+
+    pub fn pop(&mut self, samples_count: usize) -> Vec<(i16, i16)> {
+        let missing_samples = if samples_count > self.samples.len() {
+            samples_count - self.samples.len()
+        } else {
+            0
+        };
+
+        for _ in 0..missing_samples {
+            self.samples.push_back((0, 0));
+        }
+
+        self.popped_samples += samples_count as u64;
+        self.samples.drain(0..samples_count).collect()
+    }
+}
+
+/// Sums samples from inputs
+fn sum_samples<'a, I: Iterator<Item = &'a InputParams>>(
+    input_samples: &HashMap<InputId, Vec<(i16, i16)>>,
+    samples_count: usize,
+    inputs: I,
+) -> Vec<(i64, i64)> {
+    let mut summed_samples = vec![(0i64, 0i64); samples_count];
+
+    for input_params in inputs {
+        let Some(input_samples) = input_samples.get(&input_params.input_id) else {
+            continue;
+        };
+        summed_samples
+            .iter_mut()
+            .zip(input_samples.iter())
+            .for_each(|(sum, s)| {
+                sum.0 += (s.0 as f64 * input_params.volume as f64) as i64;
+                sum.1 += (s.1 as f64 * input_params.volume as f64) as i64;
+            })
+    }
+
+    summed_samples
+}
+
+/// Mix input samples accordingly to provided specification.
+fn mix(
+    input_samples: &HashMap<InputId, Vec<(i16, i16)>>,
+    output_info: &OutputInfo,
+    samples_count: usize,
+) -> AudioSamples {
+    let summed_samples = sum_samples(
+        input_samples,
+        samples_count,
+        output_info.audio.inputs.iter(),
+    );
+    // Clips sample to i16 PCM range
+    let clip = |s: i64| min(max(s, i16::MIN as i64), i16::MAX as i64) as i16;
+
+    let mixed: Vec<(i16, i16)> = match output_info.audio.mixing_strategy {
+        MixingStrategy::SumClip => summed_samples
+            .into_iter()
+            .map(|(l, r)| (clip(l), clip(r)))
+            .collect(),
+        MixingStrategy::SumScale => {
+            // abs panics in debug if val = i64::MIN, but it would so many i16 samples to sum, that it'll never happen.
+            // Assumes that v is not empty (therefore unwrap is safe)
+            let max_abs =
+                |v: &Vec<(i64, i64)>| v.iter().map(|(l, r)| (l.abs().max(r.abs()))).max().unwrap();
```

Comment:
this is hard to read, extract to separate function, do not try to cramp everything into one line